tests/UnitTests/ApiConnectors/BaseApiConnectorTest.php

```diff
@@ -73,6 +83,155 @@ private function getMaxNumRetries(): int
+        $this->client->expects($this->any())
+            ->method("sendDocument")
+            ->will($this->onConsecutiveCalls(
+                $this->throwException($e),
+                $this->throwException($e),
+                $this->returnValue($response)
+            ));
+
+        $this->service->setLogger($this);
+        $this->service->sendXmlDocument($request_document);
+
+
+        self::assertCount(6, $this->logs);
+
+        [$level, $message, $context] = $this->logs[0];
+        self::assertSame(LogLevel::DEBUG, $level);
+        self::assertSame('Sending request to Twinfield.', $message);
+        self::assertSame($this->completeXml('<dimension>value</dimension>'), $context['document_xml']);
+
+        [$level, $message, $context] = $this->logs[1];
+        self::assertSame(LogLevel::INFO, $level);
+        self::assertSame("Retrying request. Reason for initial failure: {$e->getMessage()}", $message);
+        self::assertEmpty($context);
+
+        [$level, $message, $context] = $this->logs[2];
+        self::assertSame(LogLevel::DEBUG, $level);
+        self::assertSame('Sending request to Twinfield. (attempt 2)', $message);
+        self::assertSame($this->completeXml('<dimension>value</dimension>'), $context['document_xml']);
+
+        [$level, $message, $context] = $this->logs[3];
+        self::assertSame(LogLevel::INFO, $level);
+        self::assertSame("Retrying request. Reason for initial failure: {$e->getMessage()}", $message);
+        self::assertEmpty($context);
+        self::assertEmpty($context);
+
+        [$level, $message, $context] = $this->logs[4];
+        self::assertSame(LogLevel::DEBUG, $level);
+        self::assertSame('Sending request to Twinfield. (attempt 3)', $message);
+        self::assertSame($this->completeXml('<dimension>value</dimension>'), $context['document_xml']);
+
+        [$level, $message, $context] = $this->logs[5];
+        self::assertSame(LogLevel::DEBUG, $level);
+        self::assertSame('Received response from Twinfield.', $message);
+        self::assertSame($this->completeXml('<dimension result="1">value</dimension>'), $context['document_xml']);
+    }
+
+    public function testSendDocumentLogsFailures()
+    {
+        $request_document = new \DOMDocument();
+        $request_document->loadXML('<dimension>value</dimension>');
+
+        $this->connection->expects($this->any())
+            ->method("getAuthenticatedClient")
+            ->with(Services::PROCESSXML())
+            ->willReturn($this->client);
+
+        $this->client->expects($this->any())
+            ->method("sendDocument")
+            ->will($this->onConsecutiveCalls(
+                $this->throwException(new \SoapFault('Server', 'Internal server error'))
+            ));
+
+        $this->service->setLogger($this);
+        try {
+            $this->service->sendXmlDocument($request_document);
+        } catch (\PhpTwinfield\Exception $e) {
+        }
+
+        self::assertCount(2, $this->logs);
```

Comment:
Wondering why only two logs? If it fails it should have first retried three times, right?

Reply:
This test actually mocks the ProcessXmlService to not fail, but just return a response on the first request. So no need for retries cause we're testing a successful request here :) So there's 1 request and 1 response, so 2 logs.

Reply:
Aha, got it 😉